aeppl/truncation.py

```diff
@@ -0,0 +1,95 @@
+from typing import List, Optional
+
+import aesara.tensor as at
+import numpy as np
+from aesara.assert_op import Assert
+from aesara.graph.basic import Node
+from aesara.graph.fg import FunctionGraph
+from aesara.graph.opt import local_optimizer
+from aesara.scalar.basic import Clip
+from aesara.tensor.elemwise import Elemwise
+from aesara.tensor.random.op import RandomVariable
+
+from aeppl.logprob import _logcdf, _logprob
+
+
+# TODO: Add interval transform
+class CensoredRV(RandomVariable):
+    r"""A base class for censored `RandomVariable`\s."""
+
+    def __init__(self, *args, base_op, **kwargs):
+        self.base_op = base_op
+        super().__init__(*args, **kwargs)
+
+
+@_logprob.register(CensoredRV)
+def censor_logprob(op, value, *inputs, name=None, **kwargs):
+
+    *rv_params, lower_bound, upper_bound = inputs
+    logprob = _logprob(op.base_op, value, *rv_params, name=name, **kwargs)
+    logcdf = _logcdf(op.base_op, value, *rv_params, name=name, **kwargs)
+
+    # TODO: Check constant -inf
+    # TODO: Exact bound check might be problematic
+    res = at.switch(
+        at.eq(value, lower_bound),
+        logcdf,
+        at.switch(
+            at.eq(value, upper_bound),
+            at.log(1 - at.exp(logcdf)),
+            logprob,
+        ),
+    )
+    res = at.switch(
+        at.or_(at.lt(value, lower_bound), at.gt(value, upper_bound)),
+        -np.inf,
+        res,
+    )
+
+    res = Assert("lower_bound <= upper_bound")(
+        res, at.all(at.le(lower_bound, upper_bound))
+    )
+
+    if name:
+        res.name = f"{name}_censored_logprob"
+
+    return res
+
+
+@local_optimizer(tracks=[Elemwise])
+def censor_rvs(fgraph: FunctionGraph, node: Node) -> Optional[List[Node]]:
+
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return
+
+    # TODO: Allow for one sided-truncation with set_subtensor graph (x[x>ub] = ub)
```

Comment:
Maybe the one sided censoring can be "canonicalized" to a `x = clip(x, x, ub)` and `x = clip(x, lb, x)`?

Reply:
```suggestion
    # TODO: Allow for one sided-censoring with set_subtensor graph (x[x>ub] = ub)
```